src/Biogeochemistry.jl

```diff
@@ -0,0 +1,134 @@
+#####
+##### Generic fallbacks for biogeochemistry
+#####
+
+nothing_function(args...) = nothing
+
+"""
+Update tracer tendencies.
+
+Called at the end of calculate_tendencies!
+"""
+update_tendencies!(bgc, model) = nothing
+
+"""
+Update tracer tendencies.
+
+Called at the end of calculate_tendencies!
+"""
+update_biogeochemical_state!(bgc, model) = nothing
+
+"""Return the biogeochemical forcing for `val_tracer_name` when model is called."""
+abstract type AbstractBiogeochemistry end
+
+@inline (::AbstractBiogeochemistry)(i, j, k, grid, val_tracer_name, clock, fields) = zero(grid)
+
+struct NoBiogeochemistry <: AbstractBiogeochemistry end
+
+tracernames(tracers) = keys(tracers)
+tracernames(tracers::Tuple) = tracers
+
+"""Ensure that `tracers` contains biogeochemical tracers and `auxiliary_fields` contains biogeochemical auxiliary fields (e.g. PAR)."""
+@inline function validate_biogeochemistry!(tracers, auxiliary_fields, bgc)
+    req_tracers = required_biogeochemical_tracers(bgc)
+    
+    all(tracer ∈ tracernames(tracers) for tracer in req_tracers) ||
+        error("$(req_tracers) must be among the list of tracers to use $(typeof(bgc).name.wrapper)")
+
+    req_auxiliary_fields = required_biogeochemical_auxiliary_fields(bgc)
+    
+    all(field ∈ tracernames(auxiliary_fields) for field in req_auxiliary_fields) ||
+        error("$(req_auxiliary_fields) must be among the list of auxiliary fields to use $(typeof(bgc).name.wrapper)")
+    
+    return nothing
+end
+
+required_biogeochemical_tracers(::NoBiogeochemistry) = ()
+required_biogeochemical_auxiliary_fields(bgc::AbstractBiogeochemistry) = ()
+
+"""Sets up a tracer based biogeochemical model in a similar way to SeawaterBouyancy"""
+struct TracerBasedBiogeochemistry <: AbstractBiogeochemistry
+    biogeochemical_tracers
+    reactions
+    advection_scheme
+    sinking_velocities
+    auxiliary_fields
+end
+
+function regularize_sinking_velocities(sinking_speeds)
+    sinking_velocities = []
+    for w in values(sinking_speeds)
+        u, v, w = maybe_constant_field.((0.0, 0.0, - w))
+        push!(sinking_velocities, (; u, v, w))
+    end
+
+    return NamedTuple{keys(sinking_speeds)}(sinking_velocities)
+end
+
+# we can't use the standard `ContinuousForcing` regularisation here because it requires all the tracers to be inplace to have the correct indices
+struct ContinuousBiogeochemicalForcing
+    func
+    params
+    field_dependencies
+end
+
+DiscreteBiogeochemicalForcing = DiscreteForcing
+
+ContinuousBiogeochemicalForcing(func; parameters=nothing, field_dependencies=()) = ContinuousBiogeochemicalForcing(func, parameters, field_dependencies)
+
+function BiogeochemicalForcing(func; parameters=nothing, discrete_form=false, field_dependencies=())
+    if discrete_form
+        return DiscreteBiogeochemicalForcing(func, parameters)
+    else
+        return ContinuousBiogeochemicalForcing(func, parameters, field_dependencies)
+    end
+end
```

Comment:
Do we need to re-invent `DiscreteForcing` and `ContinuousForcing`, or can we use them as is?

Reply:
`DiscreteBiogeochemicalForcing` is just an alias for `DiscreteForcing` but we definitly can't use `ContnuousForcing` because it makes interpolation operators dependent on the index of tracers so the full model needs to be setup before you can create the forcing which doesn't work here.

Reply:
We can use the same procedure / code that we use for ordinary `forcing`, I think